features/distributed/coordination/base/src/main/java/org/opennms/features/distributed/coordination/base/AbstractDomainManager.java

```diff
@@ -0,0 +1,222 @@
+            currentRole = Role.STANDBY;
+            LOG.debug("Notifying all registrants of ", currentRole, " status");
+
+            // Handlers must not block so that this happens in a timely manner
+            for (Map.Entry<String, RoleChangeHandler> handler : roleChangeHandlers.entrySet()) {
+                try {
+                    handler.getValue().becomeStandby(getDomain());
+                } catch (Exception e) {
+                    LOG.warn("Got exception while notifying handler '", handler.getKey(), "' of role '",
+                            currentRole, "' ", e);
+                }
+            }
+        }
+    }
+
+    protected synchronized final Role getCurrentRole() {
+        return currentRole;
+    }
+
+    protected final String getDomain() {
+        return domain;
+    }
+
+    /**
+     * Implementations must handle any specific logic for dealing with a registration that adds to the map of handlers
+     * for the first time (such as connecting).
+     */
+    protected abstract void onFirstRegister();
+
+    /**
+     * Implementations must handle any specific logic for dealing with a deregistration that empties the map of handlers
+     * (such as disconnecting).
+     */
+    protected abstract void onLastDeregister();
+
+    @Override
+    public final synchronized void register(String id, RoleChangeHandler roleChangeHandler) {
+        Objects.requireNonNull(id);
+
+        if (!isAnythingRegistered()) {
+            LOG.debug("Joined election pool for domain '", getDomain(), "'");
+            onFirstRegister();
+        } else if (this.roleChangeHandlers.containsKey(id)) {
+            throw new IllegalArgumentException(id + " is already registered");
+        }
+
+        LOG.debug("Adding '", id, "' to domain '", getDomain(), "'");
+        this.roleChangeHandlers.put(id, Objects.requireNonNull(roleChangeHandler));
+
+        // If we are already active we can go ahead and trigger the callback for the first time
+        if (getCurrentRole() == Role.ACTIVE) {
+            LOG.debug("Already ACTIVE, notifying registrant");
+            roleChangeHandler.becomeActive(getDomain());
+        }
+    }
+
+    @Override
+    public final synchronized void register(String id, Consumer<String> onActive, Consumer<String> onStandby) {
+        Objects.requireNonNull(onActive);
+        Objects.requireNonNull(onStandby);
+
+        register(id, new RoleChangeHandler() {
+            @Override
+            public void becomeActive(String domain) {
+                onActive.accept(domain);
+            }
+
+            @Override
+            public void becomeStandby(String domain) {
+                onStandby.accept(domain);
+            }
+        });
+    }
+
+    @Override
+    public final synchronized void deregister(String id) {
+        if (!this.roleChangeHandlers.containsKey(Objects.requireNonNull(id))) {
+            throw new IllegalArgumentException(id + " is not registered");
```

Comment:
Just a thought: Maybe `NoSuchElementException` is better suited here?

Reply:
Yes, that makes more sense.